README.md
# NAME

Mojo::Tar - Stream your (ustar) tar files

# SYNOPSIS

## Create

    use Mojo::Tar
    my $tar = Mojo::Tar->new;

    $tar->on(adding => sub ($self, $file) {
      warn sprintf qq(Adding "%s" %sb to archive\n), $file->path, $file->size;
    });

    my $cb = $tar->files(['a.baz', 'b.foo'])->create;
    open my $fh, '>', '/path/to/my-archive.tar';
    while (length(my $chunk = $cb->())) {
      print {$fh} $chunk;
    }

## Extract

    use Mojo::Tar
    my $tar = Mojo::Tar->new;

    $tar->on(extracted => sub ($self, $file) {
      warn sprintf qq(Extracted "%s" %sb\n), $file->path, $file->size;
    });

    open my $fh, '<', '/path/to/my-archive.tar';
    while (1) {
      sysread $fh, my ($chunk), 512 or die $!;
      $tar->extract($chunk);
    }

# DESCRIPTION

[Mojo::Tar](https://metacpan.org/pod/Mojo%3A%3ATar) can create and extract [ustar](http://www.gnu.org/software/tar/manual/tar.html)
tar files as a stream. This can be useful if for example your webserver is
receiving a big tar file and you don't want to exhaust the memory while
reading it.

The [pax](http://www.opengroup.org/onlinepubs/007904975/utilities/pax.html)
tar format is not planned, but a pull request is more than welcome!

Note that this module is currently EXPERIMENTAL, but the API will only change
if major design issues is discovered.

# EVENTS

## added

    $tar->on(added => sub ($tar, $file) { ... });

Emitted after the callback from ["create"](#create) has returned all the content of the `$file`.

## adding

    $tar->on(adding => sub ($tar, $file) { ... });

Emitted right before the callback from ["create"](#create) returns the tar header for the
`$file`.

## extracted

    $tar->on(extracted => sub ($tar, $file) { ... });

Emitted when ["extract"](#extract) has read the complete content of the file.

## extracting

    $tar->on(extracting => sub ($tar, $file) { ... });

Emitted when ["extract"](#extract) has read the tar header for a [Mojo::Tar::File](https://metacpan.org/pod/Mojo%3A%3ATar%3A%3AFile). This
event can be used to set the ["asset" in Mojo::Tar::File](https://metacpan.org/pod/Mojo%3A%3ATar%3A%3AFile#asset) to something else than a
temp file.

# ATTRIBUTES

## files

    $tar = $tar->files(Mojo::Collection->new('a.file', ...)]);
    $tar = $tar->files([Mojo::File->new]);
    $tar = $tar->files([Mojo::Tar::File->new, ...]);
    $collection = $tar->files;

This attribute holds a [Mojo::Collection](https://metacpan.org/pod/Mojo%3A%3ACollection) of [Mojo::Tar::File](https://metacpan.org/pod/Mojo%3A%3ATar%3A%3AFile) objects which
is used by either ["create"](#create) or ["extract"](#extract).

Setting this attribute will make sure each item is a [Mojo::Tar::File](https://metacpan.org/pod/Mojo%3A%3ATar%3A%3AFile) object,
even if the original list contained a [Mojo::File](https://metacpan.org/pod/Mojo%3A%3AFile) or a plain string.

## is\_complete

    $bool = $tar->is_complete;

True when the callback from ["create"](#create) has returned the whole tar-file or when
["extract"](#extract) thinks the whole tar file has been read.

Note that because of this, ["create"](#create) and ["extract"](#extract) should not be called on
the same object.

# METHODS

## create

    $cb = $tar->create;

This method will take ["files"](#files) and return a callback that will return a chunk
of the tar file each time it is called, and an empty string when all files has
been processed. Example:

    while (length(my $chunk = $cb->())) {
      warn sprintf qq(Got %sb of tar data\n), length $chunk;
    }

The ["adding"](#adding) and ["added"](#added) events will be emitted for each file. In addition
["is\_complete"](#is_complete) will be set when all the ["files"](#files) has been processed.

## extract

    $tar = $tar->extract($bytes);

Used to parse `$bytes` and turn the information into [Mojo::Tar::File](https://metacpan.org/pod/Mojo%3A%3ATar%3A%3AFile)
objects which are emitted as ["extracting"](#extracting) and ["extracted"](#extracted) events.

## looks\_like\_tar

    $bool = $tar->looks_like_tar($bytes);

Returns true if [Mojo::Tar](https://metacpan.org/pod/Mojo%3A%3ATar) thinks `$bytes` looks like the beginning of a
tar stream. Currently this checks if `$bytes` is at least 512 bytes long and
the checksum value in the tar header is correct.

## new

    $tar = Mojo::Tar->new(\%attrs);
    $tar = Mojo::Tar->new(%attrs);

Used to create a new [Mojo::Tar](https://metacpan.org/pod/Mojo%3A%3ATar) object. ["files"](#files) will be normalized.

# AUTHOR

Jan Henning Thorsen

# COPYRIGHT AND LICENSE

Copyright (C) Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

# SEE ALSO

[Archive::Tar](https://metacpan.org/pod/Archive%3A%3ATar)


Changes
Revision history for perl distribution Mojo-Tar

0.01 Not Released
 - Started project


Makefile.PL
use 5.016;
use strict;
use warnings;
use utf8;
use ExtUtils::MakeMaker;

my $GITHUB_URL = 'https://github.com/jhthorsen/mojo-tar';

my %WriteMakefileArgs = (
  AUTHOR        => 'Jan Henning Thorsen <jhthorsen@cpan.org>',
  LICENSE       => 'artistic_2',
  NAME          => 'Mojo::Tar',
  ABSTRACT_FROM => 'lib/Mojo/Tar.pm',
  VERSION_FROM  => 'lib/Mojo/Tar.pm',
  EXE_FILES     => [],
  TEST_REQUIRES => {'Test2::V0'   => '0.000145'},
  PREREQ_PM     => {'Mojolicious' => '9.00'},
  META_MERGE    => {
    'dynamic_config' => 0,
    'meta-spec'      => {version   => 2},
    'no_index'       => {directory => [qw(examples t)]},
    'prereqs'        => {runtime   => {requires => {perl => '5.016'}}},
    'resources'      => {
      bugtracker => {web => "$GITHUB_URL/issues"},
      homepage   => $GITHUB_URL,
      license    => ['http://www.opensource.org/licenses/artistic-license-2.0'],
      repository => {type => 'git', url => "$GITHUB_URL.git", web => $GITHUB_URL},
      x_IRC => {url => 'irc://irc.libera.chat/#convos', web => 'https://web.libera.chat/#convos'},
    },
    'x_contributors' => [],
  },
  test => {TESTS => (-e 'META.yml' ? 't/*.t' : 't/*.t xt/*.t')},
);

unless (eval { ExtUtils::MakeMaker->VERSION('6.63_03') }) {
  my $test_requires = delete $WriteMakefileArgs{TEST_REQUIRES};
  @{$WriteMakefileArgs{PREREQ_PM}}{keys %$test_requires} = values %$test_requires;
}

WriteMakefile(%WriteMakefileArgs);


lib/Mojo/Tar/File.pm
package Mojo::Tar::File;
use Mojo::Base -base, -signatures;

use Carp         qw(croak);
use Exporter     qw(import);
use Mojo::File   ();
use Scalar::Util qw(blessed);

use constant DEBUG => !!$ENV{MOJO_TAR_DEBUG};

our $GID = $( =~ /(\d+)/ && int($1) || 0;
our ($PACK_FORMAT, @EXPORT);

BEGIN {
  $PACK_FORMAT = q(
    a100 # pos=0   name=name      desc=file name (chars)
    a8   # pos=100 name=mode      desc=file mode (octal)
    a8   # pos=108 name=uid       desc=uid (octal)
    a8   # pos=116 name=gid       desc=gid (octal)
    a12  # pos=124 name=size      desc=size (octal)
    a12  # pos=136 name=mtime     desc=mtime (octal)
    a8   # pos=148 name=checksum  desc=checksum (octal)
    a1   # pos=156 name=type      desc=type
    a100 # pos=157 name=symlink   desc=file symlink destination (chars)
    A6   # pos=257 name=ustar     desc=ustar
    a2   # pos=263 name=ustar_ver desc=ustar version (00)
    a32  # pos=265 name=owner     desc=owner user name (chars)
    a32  # pos=297 name=group     desc=owner group name (chars)
    a8   # pos=329 name=dev_major desc=device major number
    a8   # pos=337 name=dev_minor desc=device minor number
    a155 # pos=345 name=prefix    desc=file name prefix
    a12  # pos=500 name=padding   desc=padding (\0)
  );

  # Generate constants:
  # TAR_USTAR_NAME_LEN       TAR_USTAR_NAME_POS
  # TAR_USTAR_MODE_LEN       TAR_USTAR_MODE_POS
  # TAR_USTAR_UID_LEN        TAR_USTAR_UID_POS
  # TAR_USTAR_GID_LEN        TAR_USTAR_GID_POS
  # TAR_USTAR_SIZE_LEN       TAR_USTAR_SIZE_POS
  # TAR_USTAR_MTIME_LEN      TAR_USTAR_MTIME_POS
  # TAR_USTAR_CHECKSUM_LEN   TAR_USTAR_CHECKSUM_POS
  # TAR_USTAR_TYPE_LEN       TAR_USTAR_TYPE_POS
  # TAR_USTAR_SYMLINK_LEN    TAR_USTAR_SYMLINK_POS
  # TAR_USTAR_USTAR_LEN      TAR_USTAR_USTAR_POS
  # TAR_USTAR_USTAR_VER_LEN  TAR_USTAR_USTAR_VER_POS
  # TAR_USTAR_OWNER_LEN      TAR_USTAR_OWNER_POS
  # TAR_USTAR_GROUP_LEN      TAR_USTAR_GROUP_POS
  # TAR_USTAR_DEV_MAJOR_LEN  TAR_USTAR_DEV_MAJOR_POS
  # TAR_USTAR_DEV_MINOR_LEN  TAR_USTAR_DEV_MINOR_POS
  # TAR_USTAR_PREFIX_LEN     TAR_USTAR_PREFIX_POS
  # TAR_USTAR_PADDING_LEN    TAR_USTAR_PADDING_POS
  for my $line (split /\n/, $PACK_FORMAT) {
    my ($len, $pos, $name) = $line =~ /(\d+)\W+pos=(\d+)\W+name=(\w+)/ or next;

    my $const = uc "TAR_USTAR_${name}_LEN";
    constant->import($const => $len);
    push @EXPORT, $const;

    $const = uc "TAR_USTAR_${name}_POS";
    constant->import($const => $pos);
    push @EXPORT, $const;
  }
}

has asset => sub ($self) {Mojo::File::tempfile};
has checksum =>
  sub ($self) { substr $self->to_header, TAR_USTAR_CHECKSUM_POS, TAR_USTAR_CHECKSUM_LEN };
has dev_major   => '';
has dev_minor   => '';
has gid         => sub ($self) { $self->_stat('gid')  || $GID };
has group       => sub ($self) { getgrgid($self->gid) || '' };
has is_complete => sub ($self) { $self->_stat('size') == $self->size ? 1 : 0 };
has mode        => sub ($self) { ($self->_stat('mode') || 0) & 0777 };
has mtime       => sub ($self) { $self->_stat('mtime')   || time };
has owner       => sub ($self) { getpwuid($self->uid)    || '' };
has path        => sub ($self) { $self->asset->to_string || '' };
has size        => sub ($self) { $self->_stat('size')    || 0 };
has symlink     => '';
has type        => sub ($self) { $self->_build_type };
has uid         => sub ($self) { $self->_stat('uid') || $( };

sub add_block ($self, $block) {
  return $self unless $self->type eq 0;

  $self->{bytes_added} //= 0;
  my $chunk = substr $block, 0, $self->size - $self->{bytes_added};
  $self->{bytes_added} += length $chunk;
  croak 'File size is out of range' if $self->{bytes_added} > $self->size;

  my $handle = $self->{add_block_handle} //= $self->asset->open('>');
  ($handle->syswrite($chunk) // -1) == length $chunk or croak "Can't write to asset: $!";
  $self->is_complete(1)->_cleanup if $self->{bytes_added} == $self->size;

  warn sprintf "[tar:add_block] chunk=%s/%s size=%s/%s is_complete=%s path=%s\n", length($chunk),
    length($block), $self->{bytes_added}, $self->size, $self->is_complete, $self->path
    if DEBUG;

  return $self;
}

sub from_header ($self, $header) {
  my @fields   = unpack $PACK_FORMAT, $header;
  my $checksum = $self->_checksum($header);

  my ($prefix, $path) = map { _trim_nul($fields[$_]) } 15, 0;
  $path = Mojo::File->new($prefix, $path)->to_string if length $prefix;

  $self->path($path);
  $self->mode(_from_oct($fields[1]));
  $self->uid(_from_oct($fields[2]));
  $self->gid(_from_oct($fields[3]));
  $self->size(_from_oct($fields[4]));
  $self->mtime(_from_oct($fields[5]));
  $self->checksum($checksum eq $fields[6] =~ s/\0\s$//r ? $checksum : '');
  $self->type($fields[7] eq "\0"                        ? '0'       : $fields[7]);
  $self->symlink(_trim_nul($fields[8]));
  $self->owner(_trim_nul($fields[11]));
  $self->group(_trim_nul($fields[12]));
  $self->dev_major($fields[13]);
  $self->dev_minor($fields[14]);

  warn sprintf
    "[tar:from_header] path=%s mode=%s uid=%s gid=%s size=%s mtime=%s checksum=%s type=%s symlink=%s owner=%s group=%s\n",
    map { $self->$_ } qw(path mode uid gid size mtime checksum type symlink owner group)
    if DEBUG;

  return $self;
}

sub new_from_path ($class, $path) {
  return $class->new->asset(Mojo::File->new("$path"))->path("$path");
}

sub to_header ($self) {
  my ($name, $prefix) = (Mojo::File->new($self->path), '');
  ($name, $prefix) = ($name->basename, $name->dirname->to_string) if length($name) > 100;
  croak qq(path "@{[$self->path]}" is too long) if length($name) > 100 or length($prefix) > 155;

  my $header = pack $PACK_FORMAT, $name,    # 0
    sprintf('%06o ',  $self->mode),         # 1
    sprintf('%06o ',  $self->uid),          # 2
    sprintf('%06o ',  $self->gid),          # 3
    sprintf('%011o ', $self->size),         # 4
    sprintf('%011o ', $self->mtime),        # 5
    '',                                     # 6 - checksum
    $self->type,                            # 7
    $self->symlink,                         # 8
    "ustar\0",                              # 9 - ustar
    '00',                                   # 10 - ustar version
    $self->owner,                           # 11
    $self->group,                           # 12
    sprintf('%07s', $self->dev_major),      # 13
    sprintf('%07s', $self->dev_minor),      # 14
    $prefix,                                # 15
    '';                                     # 16 - padding

  # Inject checksum
  substr $header, TAR_USTAR_CHECKSUM_POS, TAR_USTAR_CHECKSUM_LEN, $self->_checksum($header) . "\0 ";

  return $header;
}

sub _build_type ($self) {
  return '0' unless my $asset = $self->{asset};
  return '0' if -f $asset;                  # plain file
  return '1' if -l _;                       # symlink
  return '3' if -c _;                       # char dev
  return '4' if -b _;                       # block dev
  return '5' if -d _;                       # directory
  return '6' if -p _;                       # pipe
  return '8' if -s _;                       # socket
  return '2' if $asset->stat->nlink > 1;    # hard link
  return '9';                               # unknown
}

sub _checksum ($self, $header) {
  return sprintf '%06o', int unpack '%16C*', join '        ',
    substr($header, 0, TAR_USTAR_CHECKSUM_POS), substr($header, TAR_USTAR_TYPE_POS);
}

sub _cleanup ($self) {
  my $handle = delete $self->{add_block_handle};
  $handle->close if $handle;
}

sub _from_oct ($str) {
  $str =~ s/^0+//;
  $str =~ s/[\s\0]+$//;
  return length($str) ? oct $str : 0;
}

sub _stat ($self, $field) {
  return undef unless my $stat = $self->{stat} //= $self->{asset} && $self->{asset}->stat || 0;
  return $stat->$field;
}

sub _trim_nul ($str) {
  my $idx = index $str, "\0";
  return $idx == -1 ? $str : substr $str, 0, $idx;
}

sub DESTROY ($self) { $self->_cleanup }

1;

=encoding utf8

=head1 NAME

Mojo::Tar::File - A Mojo::Tar file

=head1 SYNOPSIS

  my $file = Mojo::Tar->new(path => 'some/file.txt');

  # This can be dangerous! Make sure path() does not contain ".."
  # or other dangerous path parts.
  $file->asset->move_to($file->path);

=head1 DESCRIPTION

L<Mojo::Asset::File> represents a tar file.

=head1 ATTRIBUTES

=head2 asset

  $file = $file->asset(Mojo::File->new);
  $asset = $file->asset;

Returns a L<Mojo::File> object. Defaults to L<Mojo::File/tempfile>.

This attribute is currently EXPERIMENTAL, but unlikely to change.

=head2 checksum

  $str = $file->checksum;

Holds the checksum read by L</from_header> or contains empty string if
the checksum does not match. This attribute can also be built from all the
attributes if L</from_header> was not called.

=head2 dev_major

This attribute is not supported yet. Pull request welcome!

=head2 dev_minor

This attribute is not supported yet. Pull request welcome!

=head2 gid

  $file = $file->gid(1001);
  $int = $file->gid;

The numeric representation of L</group>.

=head2 group

  $file = $file->group('users')
  $str = $file->group;

The string representation of L</gid>.

=head2 is_complete

  $bool = $file->is_complete;

Returns true if L</add_block> has added enough blocks to match L</size>.

=head2 mode

  $file = $file->mode(0644); # 0644 == 420
  $int = $file->mode;

The file mode. Note that this is 10-base, meaning C<$int> will be something
like "420" and not "644".

=head2 mtime

  $file = $file->mtime(time);
  $epoch = $file->mtime;

Epoch timestamp for this file.

=head2 owner

  $file = $file->owner('jhthorsen')
  $str = $file->owner;

The string representation of L</uid>.

=head2 path

  $file = $file->path('some/file/or/directory');
  $str = $file->path;

The path from the tar file. This is constructed with both the filename and
prefix (if any) in the ustar tar format.

=head2 size

  $file = $file->size(42);
  $int = $file->size;

The size of the file in bytes.

=head2 symlink

  $file = $file->symlink('path/for/symlink');
  $str = $file->symlink;

This attribute is not fully supported yet. Pull request welcome!

=head2 type

  $file = $file->type(5);
  $str = $file->type;

The tar file type.

This attribute is currently EXPERIMENTAL and might change from raw
representation to something more readable.

=head2 uid

  $file = $file->uid(1001);
  $int = $file->uid;

The numeric representation of L</owner>.

=head1 METHODS

=head2 add_block

  $file = $file->add_block($bytes);

Used to add a block from of bytes from the tar file to the L</asset>.

=head2 from_header

  $file = $file->from_header($bytes);

Will parse the header chunk from the tar file and set the L</ATTRIBUTES>.

=head2 new_from_path

  $file = Mojo::Tar::File->new_from_path('some/file');
  $file = Mojo::Tar::File->new_from_path(Mojo::File->new('some/file'));

Creates a new L<Mojo::Tar::File> object with L</asset> and L</path> set to
the input value. Other attributes will be lazy built.

=head2 to_header

  $bytes = $file->to_header;

Will construct a header chunk from the L</ATTRIBUTES>.

=head1 SEE ALSO

L<Mojo::Tar>.

=cut


lib/Mojo/Tar.pm
package Mojo::Tar;
use Mojo::Base 'Mojo::EventEmitter', -signatures;

use Carp qw(croak);
use Mojo::Collection;
use Mojo::Tar::File;
use Scalar::Util qw(blessed);

use constant DEBUG          => !!$ENV{MOJO_TAR_DEBUG};
use constant TAR_BLOCK_SIZE => 512;
use constant TAR_BLOCK_PAD  => "\0" x TAR_BLOCK_SIZE;

our $VERSION = '0.01';

has is_complete => 0;

sub create ($self) {
  my ($current, $handle, $header, $idx, $read) = (undef, undef, '', -1, 0);
  my $cb = sub (@) {
    unless ($current //= $self->files->[++$idx]) {
      return ''                 if $self->is_complete;
      warn "[tar:create] EOF\n" if DEBUG;
      return $self->is_complete(1)->emit('created') && TAR_BLOCK_PAD . TAR_BLOCK_PAD;
    }
    unless ($header) {
      warn "[tar:create] Adding @{[$current->asset]}\n" if DEBUG;
      $self->emit(adding => $current);
      return $header = $current->to_header;
    }

    $handle //= $current->asset->open('<');
    my $r = $handle->sysread(my $buffer, 131072) // croak "Unable to read @{[$current->asset]}: $!";
    $read += $r;
    return $buffer if length $buffer;

    warn "[tar:create] Added @{[$current->asset]} ($read)\n" if DEBUG;
    $self->emit(added => $current);
    ($current, $handle, $header) = (undef, undef, '');
    return __SUB__->();
  };

  return $self->is_complete(0) && $cb;
}

sub extract ($self, $bytes) {
  $self->{buf} .= $bytes;

  while (1) {
    last if length($self->{buf}) < 512;
    my $block = substr $self->{buf}, 0, 512, '';

    if (my $file = $self->{current}) {
      $file->add_block($block);
      $self->emit(extracted => delete $self->{current}) if $file->is_complete;
    }
    elsif ($block eq TAR_BLOCK_PAD) {
      warn "[tar:extract] Got tar pad block\n" if DEBUG;
      $self->{buf} = '';
      return $self->is_complete(1);
    }
    else {
      my $file = Mojo::Tar::File->new->from_header($block);
      push @{$self->files}, $file;
      $self->is_complete(0)->emit(extracting => $file);
      $file->size ? ($self->{current} = $file) : $self->emit(extracted => $file);
    }
  }

  return $self;
}

sub files ($self, $files = undef) {
  return $self->{files} //= Mojo::Collection->new unless $files;

  my $c = $self->{files} = Mojo::Collection->new;
  for my $file (@$files) {
    if (blessed($file) && $file->isa('Mojo::Tar::File')) {
      push @$c, $file;
    }
    else {
      push @$c, Mojo::Tar::File->new->asset(Mojo::File->new("$file"))->path("$file");
    }
  }

  return $self;
}

sub looks_like_tar ($self, $bytes) {
  state $padding = "\0" x TAR_USTAR_PADDING_LEN;
  return
      length($bytes) < TAR_BLOCK_SIZE                                          ? 0
    : substr($bytes, TAR_USTAR_PADDING_POS, TAR_USTAR_PADDING_LEN) ne $padding ? 0
    : Mojo::Tar::File->new->from_header($bytes)->checksum                      ? 1
    :                                                                            0;
}

sub new ($class, @attrs) {
  my $self = $class->SUPER::new(@attrs);
  $self->files($self->{files}) if $self->{files};
  return $self;
}

1;

=encoding utf8

=head1 NAME

Mojo::Tar - Stream your (ustar) tar files

=head1 SYNOPSIS

=head2 Create

  use Mojo::Tar
  my $tar = Mojo::Tar->new;

  $tar->on(adding => sub ($self, $file) {
    warn sprintf qq(Adding "%s" %sb to archive\n), $file->path, $file->size;
  });

  my $cb = $tar->files(['a.baz', 'b.foo'])->create;
  open my $fh, '>', '/path/to/my-archive.tar';
  while (length(my $chunk = $cb->())) {
    print {$fh} $chunk;
  }

=head2 Extract

  use Mojo::Tar
  my $tar = Mojo::Tar->new;

  $tar->on(extracted => sub ($self, $file) {
    warn sprintf qq(Extracted "%s" %sb\n), $file->path, $file->size;
  });

  open my $fh, '<', '/path/to/my-archive.tar';
  while (1) {
    sysread $fh, my ($chunk), 512 or die $!;
    $tar->extract($chunk);
  }

=head1 DESCRIPTION

L<Mojo::Tar> can create and extract L<ustar|http://www.gnu.org/software/tar/manual/tar.html>
tar files as a stream. This can be useful if for example your webserver is
receiving a big tar file and you don't want to exhaust the memory while
reading it.

The L<pax|http://www.opengroup.org/onlinepubs/007904975/utilities/pax.html>
tar format is not planned, but a pull request is more than welcome!

=head1 EVENTS

=head2 added

  $tar->on(added => sub ($tar, $file) { ... });

Emitted after the callback from L</create> has returned all the content of the C<$file>.

=head2 adding

  $tar->on(adding => sub ($tar, $file) { ... });

Emitted right before the callback from L</create> returns the tar header for the
C<$file>.

=head2 created

  $tar->on(created => sub ($tar, @) { ... });

Emitted right before the callback from L</create> returns the last chunk of the tar
file.

=head2 extracted

  $tar->on(extracted => sub ($tar, $file) { ... });

Emitted when L</extract> has read the complete content of the file.

=head2 extracting

  $tar->on(extracting => sub ($tar, $file) { ... });

Emitted when L</extract> has read the tar header for a L<Mojo::Tar::File>. This
event can be used to set the L<Mojo::Tar::File/asset> to something else than a
temp file.

=head1 ATTRIBUTES

=head2 files

  $tar = $tar->files(Mojo::Collection->new('a.file', ...)]);
  $tar = $tar->files([Mojo::File->new]);
  $tar = $tar->files([Mojo::Tar::File->new, ...]);
  $collection = $tar->files;

This attribute holds a L<Mojo::Collection> of L<Mojo::Tar::File> objects which
is used by either L</create> or L</extract>.

Setting this attribute will make sure each item is a L<Mojo::Tar::File> object,
even if the original list contained a L<Mojo::File> or a plain string.

=head2 is_complete

  $bool = $tar->is_complete;

True when the callback from L</create> has returned the whole tar-file or when
L</extract> thinks the whole tar file has been read.

Note that because of this, L</create> and L</extract> should not be called on
the same object.

=head1 METHODS

=head2 create

  $cb = $tar->create;

This method will take L</files> and return a callback that will return a chunk
of the tar file each time it is called, and an empty string when all files has
been processed. Example:

  while (length(my $chunk = $cb->())) {
    warn sprintf qq(Got %sb of tar data\n), length $chunk;
  }

The L</adding> and L</added> events will be emitted for each file and the
L</created> event will be emitted at the very end. In addition L</is_complete>
will also be set right before L</created> gets emitted.

=head2 extract

  $tar = $tar->extract($bytes);

Used to parse C<$bytes> and turn the information into L<Mojo::Tar::File>
objects which are emitted as L</extracting> and L</extracted> events.

=head2 looks_like_tar

  $bool = $tar->looks_like_tar($bytes);

Returns true if L<Mojo::Tar> thinks C<$bytes> looks like the beginning of a
tar stream. Currently this checks if C<$bytes> is at least 512 bytes long and
the checksum value in the tar header is correct.

=head2 new

  $tar = Mojo::Tar->new(\%attrs);
  $tar = Mojo::Tar->new(%attrs);

Used to create a new L<Mojo::Tar> object. L</files> will be normalized.

=head1 AUTHOR

Jan Henning Thorsen

=head1 COPYRIGHT AND LICENSE

Copyright (C) Jan Henning Thorsen

This program is free software, you can redistribute it and/or modify it under
the terms of the Artistic License version 2.0.

=head1 SEE ALSO

L<Archive::Tar>

=cut
